File: python/research_agent.py
"""Small Python companion to the TypeScript Agents SDK example."""

from __future__ import annotations

import asyncio
from pathlib import Path

from agents import Agent, Runner, function_tool

NOTES_DIR = Path(__file__).parent.parent / "data" / "notes"


@function_tool
def search_notes(query: str) -> str:
    """Search local Markdown notes and return matching passages with source paths."""
    words = {word.lower() for word in query.split() if len(word) > 1}
    matches: list[str] = []
    for note in NOTES_DIR.glob("*.md"):
        content = note.read_text(encoding="utf-8")
        if any(word in content.lower() for word in words):
            matches.append(f"SOURCE: data/notes/{note.name}\n{content[:600]}")
    return "\n\n".join(matches[:5]) or "NO_EVIDENCE"


research_agent = Agent(
    name="Python research assistant",
    instructions=(
        "Answer questions about AI Agent learning. Search notes first, cite each source, "
        "and explicitly say when no evidence was found."
    ),
    tools=[search_notes],
)


async def answer(question: str) -> str:
    if len(question.strip()) < 2:
        raise ValueError("question must contain at least two characters")
    result = await Runner.run(research_agent, question, max_turns=8)
    return str(result.final_output)


if __name__ == "__main__":
    import sys

    user_question = " ".join(sys.argv[1:]) or "MCP 是什么？"
    print(asyncio.run(answer(user_question)))

File: src/agent/research-agent.ts
import { Agent, run, tool } from "@openai/agents";
import { z } from "zod";

import { searchNotes } from "../core/notes.js";
import "./network.js";

const noteSearch = tool({
  name: "search_notes",
  description:
    "Search the learner's local Markdown notes. Use this before making claims about saved material.",
  parameters: z.object({
    query: z.string().min(2).max(300),
    limit: z.number().int().min(1).max(8).default(5),
  }),
  async execute({ query, limit }) {
    return searchNotes(query, undefined, limit);
  },
});

const outlineAgent = new Agent({
  name: "Research planner",
  instructions:
    "Turn the question into a short research outline. Identify missing evidence. Do not invent sources.",
});

export const researchAgent = new Agent({
  name: "AI learning research assistant",
  instructions: [
    "Help a senior frontend developer learn AI agent engineering.",
    "Search local notes before answering questions about the curriculum.",
    "Cite local evidence as [title](source).",
    "Clearly label assumptions and say when evidence is missing.",
    "Keep the answer actionable and under 800 Chinese characters unless asked for detail.",
  ].join("\n"),
  tools: [noteSearch],
  handoffs: [outlineAgent],
});

export async function answerResearchQuestion(question: string): Promise<string> {
  if (question.trim().length < 2) {
    throw new Error("Question must contain at least two characters");
  }

  const result = await run(researchAgent, question, { maxTurns: 8 });
  if (!result.finalOutput) {
    throw new Error("Agent completed without a final answer");
  }
  return result.finalOutput;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  const question = process.argv.slice(2).join(" ") || "本周应该学习什么？";
  answerResearchQuestion(question)
    .then(console.log)
    .catch((error: unknown) => {
      console.error(error instanceof Error ? error.message : error);
      process.exitCode = 1;
    });
}
